### PR TITLE
BUG: Enforce parse_dates as bool when scalar

### LINKED ISSUE #5636: read_csv not correctly parsing dates when parse_dates is string and index_col not set
With:
```
csv="""A,B,C
1,2,2003-11-1
"""
```
These all work as expected:
```
In [40]: pd.read_csv(StringIO(csv), parse_dates="C",index_col="C").index[0]

Out[40]: Timestamp('2003-11-01 00:00:00', tz=None)

In [41]: pd.read_csv(StringIO(csv), parse_dates=["C"],index_col="C").index[0]

Out[41]: Timestamp('2003-11-01 00:00:00', tz=None)

In [42]: pd.read_csv(StringIO(csv), parse_dates=["C"]).C[0]

Out[42]: Timestamp('2003-11-01 00:00:00', tz=None)
```

but this does not parse the string:

```
In [39]: pd.read_csv(StringIO(csv), parse_dates="C",).C[0]

Out[39]: '2003-11-1'
```

### PR DESCRIPTION
**Pull request by gfyoung**
Title is self-explanatory.  Closes #5636.

### CODE REVIEW

File: pandas/io/tests/test_parsers.py [hunk 1]
```diff
@@ -1036,7 +1036,7 @@ def test_parse_dates_string(self):
 20090103,c,4,5
 """
         rs = self.read_csv(
```
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: can you move all of the parse_dates testing to a separate class (make it a base class, kind of like ``ParserTests`` e.g. it inherits from ``object`` and have ``ParserTests`` inherit from it. Ideally we would group tests like this to keep them together. Can do this in another PR if you want to organize things even more.

Comment (jreback) [inline pandas/io/tests/test_parsers.py]: in any event move the ``parse_dates`` tests all together at least.

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -825,6 +825,22 @@ def _validate_usecols_arg(usecols):
     return usecols
 
 
+def _validate_parse_dates_arg(parse_dates):
```
Comment (jreback) [inline pandas/io/parsers.py]: you are adding new code but not checking for existing code, e.g. see like 298. Further let's go ahead and validate all of the cases (and test whatever is not tested), e.g. list/dict are valid (but need mainly to tests what is not accepted)

Comment (jreback) [inline pandas/io/parsers.py]: e.g. can accept list-of-(lists, ints, strings), or dict.

Comment (gfyoung) [inline pandas/io/parsers.py]: Existing code and logic does not validate IINM.  For example, line 298 is about overriding `parse_dates` if it is a boolean and `date_parser` is passed.

Comment (gfyoung) [inline pandas/io/parsers.py]: On a separate note, to what extent should this validation go?  Is it just type checking?  Or are we also validating element types?  Also, are mixed types allowed?  I presume a mixture of `int` and `string` is not, but what about `int` and `list` for example?

File: pandas/io/tests/test_parsers.py (hunk 1 above)
Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: I moved what I think are all of the _pure_ `parse_dates` test together.

File: pandas/io/parsers.py (hunk 2 above)
Comment (jreback) [inline pandas/io/parsers.py]: so you are check scalars, just check that it is a ``list`` or ``dict``, that would be good enough for now. prob don't have a tests with passing something that is not a list/doc/scalar

Comment (gfyoung) [inline pandas/io/parsers.py]: Done.

File: doc/source/whatsnew/v0.18.1.txt [hunk 3]
```diff
@@ -129,7 +129,7 @@ API changes
 
 - ``Period`` and ``PeriodIndex`` now raises ``IncompatibleFrequency`` error which inherits ``ValueError`` rather than raw ``ValueError`` (:issue:`12615`)
 
-
+- ``read_csv`` will now raise a ``TypeError`` if ``parse_dates`` is neither a boolean, list, or dictionary (:issue:`5636`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: what did it do before?

Comment (gfyoung) [inline doc/source/whatsnew/v0.18.1.txt]: nothing

**Comment by jreback** [comment]
ok, lgtm. ping when greeen.

**Comment by gfyoung** [comment]
@jreback : Travis finally got the OSX build done, and everything is green!  Ready to merge.
